File: 2846-robot-collisions/2846-robot-collisions.py
class Robot:
    def __init__(self, pos, health, direction, index):
        self.position = pos
        self.health = health
        self.direction = direction
        self.index = index

class Solution(object):
    def survivedRobotsHealths(self, positions, healths, directions):
        n = len(positions)
        robots = []
        for i in range(n):
            curr = Robot(positions[i], healths[i], directions[i], i)
            robots.append(curr)

        robots.sort(key=lambda robot: robot.position)

        stack = []
        for robot in robots:
            if robot.direction == 'R':
                stack.append(robot)
                continue

            flag = False
            while stack and stack[-1].health <= robot.health and stack[-1].direction == 'R':
                if stack[-1].health == robot.health:
                    stack.pop()
                    flag = True
                    break
                robot.health -= 1
                stack.pop()

            if not flag and stack and stack[-1].direction == 'R' and stack[-1].health > robot.health:
                stack[-1].health -= 1
                flag = True

            if not flag:
                stack.append(robot)

        stack.sort(key=lambda robot: robot.index)

        ans = [robot.health for robot in stack]
        return ans
        

File: 2846-robot-collisions/2846-robot-collisions.cpp
class Robot {
public:
    int position;
    int health;
    char direction;
    int index;

    Robot(int pos, int h, char dir, int idx) {
        position = pos;
        health = h;
        direction = dir;
        index = idx;
    }
};

class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        int n = positions.size();
        vector<Robot> robots;
        for(int i = 0; i<n; i++) {
            Robot curr = Robot(positions[i], healths[i], directions[i], i);
            robots.push_back(curr);
        }

        sort(robots.begin(), robots.end(), [](const Robot &a, const Robot &b) {
            return a.position < b.position;
        });

        vector<Robot> stack;
        for(auto robot : robots) {
            if(robot.direction == 'R') {
                stack.push_back(robot);
                continue;
            }

            bool flag = false;
            while(!stack.empty() && stack.back().health <= robot.health && stack.back().direction == 'R') {
                if(stack.back().health == robot.health) {
                    stack.pop_back();
                    flag = true;
                    break;
                }
                robot.health--;
                stack.pop_back();
            }

            if(!flag && !stack.empty() && stack.back().direction == 'R' && stack.back().health > robot.health) {
                stack.back().health--;
                flag = true;
            }

            if(!flag) stack.push_back(robot);
        }

        sort(stack.begin(), stack.end(), [](const Robot &a, const Robot &b) {
            return a.index < b.index;
        });

        vector<int> ans;
        for(auto robot : stack) {
            ans.push_back(robot.health);
        }

        return ans;
    }
};